### ISSUE TITLE
How do I block the backprop after k steps?

### ISSUE THREAD

**Issue by ihexx** [facebookresearch/higher#9]
Hi there, great library, thanks for open sourcing it 👍 
In my problem, I only need to backpropagate through K steps, and I'm trying to figure out how to 'reset' the functional models and differentiable optimizers at that point while saving state.
```python
for i in range(num_steps): # training loop
    with  higher.innerloop_ctx(model,optim) as (fmodel,diffoptim):
        for k in range(2): # inner loop i need to backprop through
            ....#get losses
            train_step(fmodel,loss,diffoptim)
    ... # do meta updates
    ... # block backprop here somehow?
```
I'm trying to avoid the cost of backpropagating through `num_steps` steps instead of just `k` which is what would happen if i put the `higher.innerloop_ctx` on top of the outer loop

I figured i could put it on top of the inner loop and just recreate `fmodel` and `diffoptim` on each train step, then just copy their state back to the normal pytorch ones; `model` and `optim` after the meta update 

Copying model state was easy enough since the state dict and parameters still work like nn.Modules
But i'm stuck trying to figure out how to copy the optimizer state.

Is this the correct way to go about trying block the backprop computation in higher?
if so, how do i copy the optimizer states back to the normal torch optimizer

**Comment by ihexx** [comment]
I got around this with a hack that makes the `DifferentiableOptimizer` class copy the `.state_dict()` method from the normal pytorch optimizers, and made it populate the state dict (for adam) in a way that works for my problem, but probably doesn't in any other case.
I can now call `optim.load_state_dict(diffoptim.state_dict())` to copy the optimiser back.

[here's a gist of the edits](https://gist.github.com/ihexx/6689c641b93fc0c476f4a7181ba295a5)

I'm still curious to know what the right way of doing this is.

**Comment by ihexx** [comment]
I got around this by hacking together a `.state_dict()` method for DifferentiableOptimizer, which works in my case, but I'm pretty sure fails for higher values of k.

It allows me to call `optim.load_state_dict(diffoptim.state_dict())` and just re-create the higher models on each step

I'm still curious what the non-hacky way of doing something like this would be with higher though.
[For posterity, here's a gist of my change](https://gist.github.com/ihexx/02d54a29abcec118d66504a53b74efaa/revisions?diff=unified)

**Comment by egrefen** [comment]
If I understand correctly, what you wish to do is unroll the inner loop, take gradients with regard to meta variables where needed, but instead of "throwing away" the optimizations performed during your inner loop, to update the outer model and optimizer with the state of the inner model and optimizer at the end of an unrolled loop, is that correct?

If so, could you raise two specific issues asking for utility functions for the state of `fmodel` and `diffopt` back to `model` and `opt` at any arbitrary point? Suggestions at to the signature of these functions are welcome if you have any. Doing things with "official" functions will hopefully be more robust and less hacky, but I'll have a look at your gist when implementing them.

**Comment by ihexx** [comment]
Yeah that's exactly right. I'll open a new issue for the differentiable optimizer only since the functional models already do this. Thanks
